cryptofeed/auth/kucoin.py

```diff
@@ -13,16 +13,19 @@
 def generate_token(key_id: str, key_secret: str, key_passphrase: str, str_to_sign: str) -> dict:
     # https://docs.kucoin.com/#authentication
 
-    now = int(time.time() * 1000)
-    str_to_sign = str(now) + str_to_sign
-    signature = base64.b64encode(
-        hmac.new(key_secret.encode('utf-8'), str_to_sign.encode('utf-8'), hashlib.sha256).digest())
+    # Now required to pass timestamp with string to sign. Timestamp should exactly match header timestamp
+    now = str(int(time.time() * 1000))
+    str_to_sign = now + str_to_sign
+    signature = base64.b64encode(hmac.new(key_secret.encode('utf-8'), str_to_sign.encode('utf-8'), hashlib.sha256).digest())
+    # Passphrase must now be encrypted by key_secret
+    passphrase = base64.b64encode(hmac.new(key_secret.encode('utf-8'), key_passphrase.encode('utf-8'), hashlib.sha256).digest())
 
+    # API key version is currently 2 (whereas API version is anywhere from 1-3 ¯\_(ツ)_/¯)
     header = {
-        "KC-API-SIGN": signature.decode(),
-        "KC-API-TIMESTAMP": str(now),
         "KC-API-KEY": key_id,
-        "KC-API-PASSPHRASE": key_passphrase,
-        "KC-API-KEY-VERSION": "3"
+        "KC-API-SIGN": signature.decode(),
+        "KC-API-TIMESTAMP": now,
+        "KC-API-PASSPHRASE": passphrase.decode(),
+        "KC-API-KEY-VERSION": "2"
```

Comment:
I'm un-clear why this header needs to change? I tested the kucoin auth with the current version in the code and didn't have any issues with it

Reply:
There is no v3 of the API _**Key**_. (The API itself spans v1-3, depending on the method)

They have [only just deprecated v1](https://www.kucoin.com/news/en-announcement-regarding-the-postponement-of-kucoin-api-key-2-upgrade) so I took this bit of code to be an error.

From [the docs](https://docs.kucoin.top/#authentication): 

"KC-API-KEY-VERSION You can check the version of API key on the page of API Management 
_(All my new ones are automatically v2)_
...
For the KC-API-PASSPHRASE of the header:

For API key-V1.0, please pass requests in plaintext.
For API key-V2.0, please Specify KC-API-KEY-VERSION as 2 "
_(No mention of v3 anywhere)_

Reply:
Did you try to run with your new config? I know their docs state that the passphrase needs to be a signed key, but that didn't work for me. I wrote that function specifically for the order book endpoint, which if you check the documentation will see that v2 was deprecated, and v3 is left. It was working well when I left it.

https://docs.kucoin.com/#get-full-order-book-aggregated